core/arch/arm/kernel/mutex.c

```diff
@@ -26,212 +26,42 @@
-
-	if (handle == -1)
-		panic();
-	return handle;
-}
-
-static void mutex_handle_put(int handle)
-{
-	bool bit_was_set;
-
-	if (handle == -1) /* Uninitialized mutex */
-		return;
-
-	TEE_ASSERT(handle < MUTEX_MAX_NUMBER_OF);
-
-	bit_was_set = bit_test(mutex_handle_db, handle);
-	if (bit_was_set)
-		bit_clear(mutex_handle_db, handle);
-
-	TEE_ASSERT(bit_was_set);
-}
-
-static void mutex_check_init(struct mutex *m)
-{
-	uint32_t old_itr_status;
-
-	if (m->handle != -1)
-		return;
-
-	old_itr_status = thread_mask_exceptions(THREAD_EXCP_ALL);
-	cpu_spin_lock(&mutex_handle_db_spin_lock);
-
-	if (m->handle == -1)
-		m->handle = mutex_handle_get();
-
-	cpu_spin_unlock(&mutex_handle_db_spin_lock);
-	thread_unmask_exceptions(old_itr_status);
-}
 
 void mutex_init(struct mutex *m)
 {
 	*m = (struct mutex)MUTEX_INITIALIZER;
-	mutex_check_init(m);
-}
-
-static void mutex_wait_cmd(uint32_t cmd, int handle, uint32_t tick)
-{
-	struct tee_ta_session *sess = NULL;
-	struct teesmc32_arg *arg;
-	struct teesmc32_param *params;
-	const size_t num_params = 2;
-	paddr_t pharg = 0;
-
-	tee_ta_get_current_session(&sess);
-	if (sess)
-		tee_ta_set_current_session(NULL);
-
-	pharg = thread_rpc_alloc_arg(TEESMC32_GET_ARG_SIZE(num_params));
-
-	/*
-	 * If allocation fails, spin on the mutex, maybe there's another
-	 * thread that will release the mutex. The only other option is to
-	 * panic.
-	 */
-
-	if (!pharg)
-		goto exit;
-
-	if (!TEE_ALIGNMENT_IS_OK(pharg, struct teesmc32_arg))
-		goto exit;
-
-	if (core_pa2va(pharg, &arg))
-		goto exit;
-
-	arg->cmd = TEE_RPC_WAIT_MUTEX;
-	arg->ret = TEE_ERROR_GENERIC;
-	arg->num_params = num_params;
-	params = TEESMC32_GET_PARAMS(arg);
-	params[0].attr = TEESMC_ATTR_TYPE_VALUE_INPUT;
-	params[1].attr = TEESMC_ATTR_TYPE_VALUE_INPUT;
-	params[0].u.value.a = cmd;
-	params[1].u.value.a = handle;
-	params[1].u.value.b = tick;
-
-	thread_rpc_cmd(pharg);
-exit:
-	thread_rpc_free_arg(pharg);
-	if (sess)
-		tee_ta_set_current_session(sess);
 }
 
 void mutex_lock(struct mutex *m)
 {
-	bool did_sleep = false;
-
-	mutex_check_init(m);
-
-	while (true) {
-		enum mutex_value old_value;
-		uint32_t old_itr_status;
-		uint32_t tick = 0;
-		int handle = -1;
-
-		old_itr_status = thread_mask_exceptions(THREAD_EXCP_ALL);
-		cpu_spin_lock(&m->spin_lock);
-
-		if (did_sleep)
-			m->num_waiters--;
-
-		old_value = m->value;
-		if (old_value == MUTEX_VALUE_LOCKED) {
-			m->num_waiters++;
-			m->tick++;
-			tick = m->tick;
-			handle = m->handle;
-		} else
-			m->value = MUTEX_VALUE_LOCKED;
-
-
-		cpu_spin_unlock(&m->spin_lock);
-		thread_unmask_exceptions(old_itr_status);
-
-		if (old_value == MUTEX_VALUE_UNLOCKED)
-			return; /* We have the lock */
-
+	while (!mutex_trylock(m)) {
 		/*
 		 * Someone else is holding the lock, wait in normal world
 		 * for the lock to become available.
 		 */
 		DMSG("thread: %u sleeping", thread_get_id());
-		did_sleep = true;
-		mutex_wait_cmd(TEE_WAIT_MUTEX_SLEEP, handle, tick);
+		wq_wait(&m->wq);
```

Comment:
There's race here, when the mutex becomes unlocked before wq_wait has added its wait element to the wait queue.